types/datum.go

```diff
@@ -2520,3 +2520,12 @@ func (d Datum) EstimatedMemUsage() int64 {
 	}
 	return int64(bytesConsumed)
 }
+
+// CalcBytesHash is used to calculate the hash value of data
+func CalcBytesHash(data []byte, nr1, nr2 uint64) (uint64, uint64) {
+	for _, d := range data {
+		nr1 ^= ((nr1&63)+nr2)*(uint64(d)) + (nr1 << 8)
+		nr2 += 3
+	}
+	return nr1, nr2
+}
```

Comment:
This hash algorithm comes from MySQL.
um... I'm not sure which hash algorithm should be chosed, md5 or others ? @mjonss @
bb7133

Reply:
I tested the distribution of this, and it will have very skewed distribution (50-100+% more rows in some partitions), both for bigint and varchar (in MySQL varchar seems OK, but still skewed bigint value distribution between the partitions).

[key-part-dist.txt](https://github.com/pingcap/tidb/files/10758197/key-part-dist.txt)

Reply:
The test case above still cause skewed distribution for varchar partition filed in MySQL.
But I think there are no hash algorithm is adapted to all scenes.

Reply:
Since we will not be able to have the exact same mapping of data to partitions as in MySQL (due to different storage format for different column types etc.) we are free to choose a decent hash algorithm. I would propose using a standard, well known non-crypto hash algorithm with good "random" distribution. Some suggestions would be crc32, md5 or even newer ones like xxHash etc.

Is it possible to benchmark a few different algorithms to make sure the overhead is not too big (i.e. noticeable for a batched insert or join operation, were a lot of rows needs to evaluate their partition ID)?